index.js

```diff
@@ -797,17 +797,16 @@
   //. the type comprising every [`String`][] value except `''`.
   //.
   //. The given type must satisfy the [Monoid][] and [Setoid][] specifications.
-  var NonEmpty = UnaryType(
-    'sanctuary-def/NonEmpty',
-    functionUrl('NonEmpty'),
-    Any,
-    function(x) {
+  function NonEmpty(parent) {
+    function test(x) {
       return Z.Monoid.test(x) &&
              Z.Setoid.test(x) &&
              !Z.equals(x, Z.empty(x.constructor));
-    },
-    function(monoid) { return [monoid]; }
-  );
+    }
+    function extract(monoid) { return [monoid]; }
+    var t = UnaryTypeWithUrl('sanctuary-def/NonEmpty', parent, test, extract);
```

Comment:
Using `parent` as an *actual* parent type, like I'm doing here, seems like a good idea. Now `NonEmpty String` will (correctly) be a "refinement" of `String`.

However, as indicated by the failing unit test, it causes a value of `[1]` for `$NonEmpty($Array($String))`, for example, to generate this error message:

```text
NonEmpty (Array String)
         ^^^^^^^^^^^^^^
               1

1) 1 :: Number

The value at position 1 is not a member of ‘Array String’. 
```

Where we would want:

```text
NonEmpty (Array String)
                ^^^^^^
                  1

1) 1 :: Number

The value at position 1 is not a member of ‘String’.
```

I'm not sure why this happens. Replacing `parent` with `Any` solves the issue.